File: code/sorting/src/quick_sort/quick_sort.lua
function quicksort(t, start, endi)
	start, endi = start or 1, endi or #t
	--partition w.r.t. first element
	if(endi - start < 1) then return t end
	local pivot = start
	for i = start + 1, endi do
		if t[i] <= t[pivot] then
			if i == pivot + 1 then
				t[pivot], t[pivot + 1] = t[pivot + 1], t[pivot]
			else
				t[pivot], t[pivot + 1], t[i] = t[i], t[pivot], t[pivot + 1]
			end
			pivot = pivot + 1
		end
	end
	t = quicksort(t, start, pivot - 1)
	return quicksort(t, pivot + 1, endi)
end

File: cosmos-master/code/sorting/quick_sort/quick_sort.py
def quickSort(arr):
   quickSortHelper(arr,0,len(arr)-1)

def quickSortHelper(arr,first,last):
   if first<last:

       splitpoint = partition(arr,first,last)

       quickSortHelper(arr,first,splitpoint-1)
       quickSortHelper(arr,splitpoint+1,last)


def partition(arr,first,last):
   pivot = arr[first]

   left = first+1
   right = last

   done = False
   while not done:

       while left <= right and arr[left] <= pivot:
           left = left + 1

       while arr[right] >= pivot and right >= left:
           right = right -1

       if right < left:
           done = True
       else:
           temp = arr[left]
           arr[left] = arr[right]
           arr[right] = temp

   temp = arr[first]
   arr[first] = arr[right]
   arr[right] = temp


   return right